### PR TITLE
[SYCL] Add support of multiple devices within a context

### PR DESCRIPTION
**Pull request by dm-vodopyanov**
This patch adds support of multiple devices within a context. Programs
can be created from images or from SPIR-V binaries. Only kernels,
created using invoking kernels functions (parallel_for, single_task,
...) are supported. Kernels, created in OpenCL interoperability mode
(using sycl::program and sycl::kernel functions) are not supported.

### CODE REVIEW

File: sycl/source/detail/scheduler/commands.cpp [hunk 1]
```diff
@@ -1909,6 +1909,12 @@ cl_int ExecCGCommand::enqueueImp() {
     if (nullptr != ExecKernel->MSyclKernel) {
       assert(ExecKernel->MSyclKernel->get_info<info::kernel::context>() ==
              Context);
+      if (Context.get_devices().size() > 1) {
```
Comment (romanovvlad) [inline sycl/source/detail/scheduler/commands.cpp:1912]: Could you please move this check to the sycl::program constructor?

Comment (dm-vodopyanov) [reply sycl/source/detail/scheduler/commands.cpp]: Done, it looks like it's better to move it to ctors of program_impl class

File: sycl/test/context-with-multiple-devices/context-with-multiple-devices.cpp [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+      std::rethrow_exception(e);
+    } catch (sycl::exception const &e) {
+      std::cout << "Caught asynchronous SYCL exception:\n"
+                << e.what() << std::endl;
+    }
+  }
+}
+
+int main() {
+  std::vector DeviceList = sycl::device::get_devices();
```
Comment (v-klochkov) [inline sycl/test/context-with-multiple-devices/context-with-multiple-devices.cpp:25]: I would think this is a syntax error (when <device> is not specified). Hmm..., how/why does this work?

Comment (dm-vodopyanov) [reply sycl/test/context-with-multiple-devices/context-with-multiple-devices.cpp:25]: The compiler compiles a code with C++17 enabled by default now (it's applicable only to user code, not to DPC++ RT, it's still on C++14), so as this test is kinda user code, I thought it would be good to use CTAD feature from C++17 here.

File: sycl/source/detail/program_manager/program_manager.hpp [hunk 3]
```diff
@@ -203,7 +209,7 @@ class ProgramManager {
   // the underlying program disposed of), so the map can't be used in any way
   // other than binary image lookup with known live PiProgram as the key.
   // NOTE: access is synchronized via the MNativeProgramsMutex
-  std::unordered_map<pi::PiProgram, const RTDeviceBinaryImage *> NativePrograms;
+  std::map<pi::PiProgram, const RTDeviceBinaryImage *> NativePrograms;
```
Comment (sergey-semenov) [inline sycl/source/detail/program_manager/program_manager.hpp:212]: This change seems unrelated.

File: sycl/test/context-with-multiple-devices/context-with-multiple-devices.cpp (hunk 2 above)
Comment (keryell) [inline sycl/test/context-with-multiple-devices/context-with-multiple-devices.cpp:25]:
```suggestion
  auto DeviceList = sycl::device::get_devices();
```
is even simpler and works since C++11

File: sycl/test/context-with-multiple-devices/context-with-multiple-devices.cpp [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  DeviceList.erase(std::remove_if(DeviceList.begin(), DeviceList.end(),
+                                  [](auto Device) { return Device.is_host(); }),
+                   DeviceList.end());
+
+  sycl::context Context(DeviceList, &exceptionHandler);
+
+  std::vector<sycl::queue> QueueList;
+  for (const auto &Device : Context.get_devices()) {
+    sycl::queue Queue(Context, Device, &exceptionHandler);
+    QueueList.push_back(Queue);
```
Comment (keryell) [inline sycl/test/context-with-multiple-devices/context-with-multiple-devices.cpp:37]:
```suggestion
    QueueList.emplace_back(Context, Device, &exceptionHandler);
```

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 5]
```diff
@@ -739,68 +711,61 @@ getDeviceLibPrograms(const ContextImplPtr Context,
   // Disable all devicelib extensions requiring fp64 support if at least
   // one underlying device doesn't support cl_khr_fp64.
   bool fp64Support = true;
```
Comment (romanovvlad) [inline sycl/source/detail/program_manager/program_manager.cpp:713]:
```suggestion
```

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 6]
```diff
@@ -739,68 +711,61 @@ getDeviceLibPrograms(const ContextImplPtr Context,
...
-        get_device_info<std::string, info::device::extensions>::get(
-            Dev, Context->getPlugin());
-    fp64Support =
-        fp64Support && (DevExtList.npos != DevExtList.find("cl_khr_fp64"));
-  }
+  std::string DevExtList =
+      get_device_info<std::string, info::device::extensions>::get(
+          Device, Context->getPlugin());
+  fp64Support =
+      fp64Support && (DevExtList.npos != DevExtList.find("cl_khr_fp64"));
```
Comment (romanovvlad) [inline sycl/source/detail/program_manager/program_manager.cpp:718]:
```suggestion
  const bool fp64Support = (DevExtList.npos != DevExtList.find("cl_khr_fp64"));
```

File: sycl/source/detail/program_manager/program_manager.hpp (hunk 3 above)
Comment (dm-vodopyanov) [reply sycl/source/detail/program_manager/program_manager.hpp]: Thanks, sorry.

**Review by romanovvlad — APPROVED** [review]

**Review by sergey-semenov — APPROVED** [review]
